### PR TITLE
Prune epoch doesn't remove *all* micro blocks 

### LINKED ISSUE #1197: Prune epoch doesn't remove *all* micro blocks
`prune_epoch` is supposed to remove all micro blocks of an epoch.
To this end, it iterates over the block numbers and retrieves the blocks from the height index.

However, when there are forks, there could be duplicates at this height (see `get_blocks_at` for how to handle these).
Currently, we disregard such blocks.

Instead, we should iterate over all duplicates for a given height just as in the `get_blocks_at` function.

### PR DESCRIPTION
**Pull request by cud4m**
- Added test for MicroBlock prune
- Added get_block_hashes function. 
- Refactor prune_epoch function to use new get_block_hashes

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.

### CODE REVIEW

File: blockchain/src/chain_store.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
+use std::hash::Hash;
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]: Looks like this import is not needed

File: blockchain/src/chain_store.rs [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
 use nimiq_account::Receipts;
 use nimiq_block::Block;
 use nimiq_database::cursor::{ReadCursor, WriteCursor};
-use nimiq_database::{
-    Database, DatabaseFlags, Environment, ReadTransaction, Transaction, WriteTransaction,
-};
+use nimiq_database::{cursor, Database, DatabaseFlags, Environment, ReadTransaction, Transaction,
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]: Looks like the `cursor` import is not needed

File: blockchain/tests/mod.rs [hunk 3]
```diff
@@ -2,13 +2,13 @@ use parking_lot::RwLock;
 use std::sync::Arc;
 use tokio_stream::wrappers::BroadcastStream;
 use tokio_stream::StreamExt;
-
 use nimiq_block::Block;
 use nimiq_block_production::BlockProducer;
-use nimiq_blockchain::PushResult;
+use nimiq_blockchain::{ChainInfo, PushResult};
```
Comment (jsdanielh) [inline blockchain/tests/mod.rs]: Looks like the `ChainInfo` import is not needed

File: blockchain/tests/mod.rs [hunk 4]
```diff
@@ -2,13 +2,13 @@ use parking_lot::RwLock;
...
 use nimiq_blockchain::{AbstractBlockchain, Blockchain, BlockchainConfig};
 use nimiq_database::volatile::VolatileEnvironment;
 use nimiq_genesis::NetworkId;
+use nimiq_hash::Blake2bHash;
```
Comment (jsdanielh) [inline blockchain/tests/mod.rs]: Looks like this import is not needed

**Review by jsdanielh — COMMENTED** [review]
Also, could you run `cargo fmt`?. Looks like it isn't passing this CI job.

File: blockchain/src/chain_store.rs [hunk 5]
```diff
@@ -272,6 +272,35 @@ impl ChainStore {
...
+            }
+        };
+
+        let mut hashes: Vec<Blake2bHash> = Vec::new();
+        // Seek the hash of the first block at the given height and add it to our hashes vector
+        let mut cursor = txn.cursor(&self.height_idx);
+        let mut hash = cursor.seek_key::<u32, Blake2bHash>(&block_height)
+            .map(|hash| (0u32, hash));
+
+        //Get any other block hash at the given height and add them to our hashes vector
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]:
```suggestion
        // Get any other block hash at the given height and add them to our hashes vector
```

File: blockchain/src/chain_store.rs [hunk 6]
```diff
@@ -512,16 +540,19 @@ impl ChainStore {
...
-                    txn.remove(&self.chain_db, &hash);
-                    txn.remove(&self.block_db, &hash);
-                    txn.remove_item(&self.height_idx, &height, &hash);
+            if let Ok(hashes) = self.get_block_hashes_at(height,
+                                                         Some(txn)) {
+                for hash in hashes {
+                    let chain_info: ChainInfo = txn
+                        .get(&self.chain_db, &hash)
+                        .expect("Corrupted store: ChainInfo referenced from index not found");
+                     // If we detect a block whose prunable flag is set to false, we don't prune it
```
Comment (jsdanielh) [inline blockchain/src/chain_store.rs]:
```suggestion
                    // If we detect a block whose prunable flag is set to false, we don't prune it
```

File: blockchain/src/chain_store.rs (hunk 1 above)
Comment (cud4m) [reply blockchain/src/chain_store.rs]: Done! I'll be more careful from now on

**Comment by cud4m** [comment]
The conflict is because of the import. There's really no need to have it split in two lines

File: blockchain/src/chain_store.rs [hunk 7]
```diff
@@ -272,6 +272,36 @@ impl ChainStore {
         }
     }
 
+    pub fn get_block_hashes_at(
+        &self,
+        block_height: u32,
+        txn_option: Option<&Transaction>,
+    ) -> Result<Vec<Blake2bHash>, BlockchainError> {
```
Comment (paberr) [inline blockchain/src/chain_store.rs]: As far as I can see, the function actually cannot return an error. So, a return type of `Vec<Blake2bHash>` would be sufficient.

File: blockchain/tests/mod.rs [hunk 8]
```diff
@@ -1,19 +1,17 @@
...
 use nimiq_genesis::NetworkId;
 use nimiq_primitives::policy::Policy;
 use nimiq_test_log::test;
 use nimiq_test_utils::block_production::TemporaryBlockProducer;
 use nimiq_test_utils::blockchain::{signing_key, voting_key};
 use nimiq_utils::time::OffsetTime;
+use parking_lot::RwLock;
+use std::sync::Arc;
+use tokio_stream::wrappers::BroadcastStream;
+use tokio_stream::StreamExt;
```
Comment (paberr) [inline blockchain/tests/mod.rs]: The previous ordering of imports was correct.
We start with a block of external dependencies, then a block of nimiq dependencies, then a block of crate dependencies.

File: blockchain/tests/mod.rs [hunk 9]
```diff
@@ -139,6 +137,106 @@ fn it_can_push_consecutive_view_changes() {
...
+        .is_ok());
+    assert!(bc_read
+        .chain_store
+        .get_chain_info(&micro_block2.hash(), false, None,)
+        .is_ok());
+    assert!(bc_read
+        .chain_store
+        .get_chain_info(&micro_block3.hash(), false, None,)
+        .is_ok());
+    assert_eq!(blockchain.read().block_number(), 1);
```
Comment (paberr) [inline blockchain/tests/mod.rs]:
```suggestion
    assert_eq!(bc_read.block_number(), 1);
```

File: blockchain/src/chain_store.rs (hunk 7 above)
Comment (cud4m) [reply blockchain/src/chain_store.rs]: Yes, you are completely right. Modified in accordance

File: blockchain/tests/mod.rs (hunk 9 above)
Comment (cud4m) [reply blockchain/tests/mod.rs]: Modified

File: blockchain/tests/mod.rs [hunk 10]
```diff
@@ -139,6 +138,106 @@ fn it_can_push_consecutive_view_changes() {
...
+        .get_chain_info(&micro_block2.hash(), false, None,)
+        .is_ok());
+    assert!(bc_read
+        .chain_store
+        .get_chain_info(&micro_block3.hash(), false, None,)
+        .is_ok());
+    assert_eq!(bc_read.block_number(), 1);
+
+    let mut txs = bc_read.write_transaction();
+    // prune the 3 created MicroBlocks
```
Comment (paberr) [inline blockchain/tests/mod.rs]:
```suggestion
    // Prune the 3 created MicroBlocks
```

File: blockchain/tests/mod.rs [hunk 11]
```diff
@@ -139,6 +138,106 @@ fn it_can_push_consecutive_view_changes() {
...
+    bc_read.chain_store.prune_epoch(1, &mut txs);
+
+    // check that they no longer exist
```
Comment (paberr) [inline blockchain/tests/mod.rs]:
```suggestion
    // Check that they no longer exist
```

File: blockchain/src/chain_store.rs [hunk 12]
```diff
@@ -298,7 +328,6 @@ impl ChainStore {
         // Iterate until we find all blocks at the same height.
         while let Ok(block) = self.get_block(&block_hash, include_body, Some(txn)) {
             blocks.push(block);
-
```
Comment (paberr) [inline blockchain/src/chain_store.rs:306]: Readd new line

**Review by paberr — COMMENTED** [review]
Needs rebase + small changes, then it's ready to be merged

**Pull request merged**